control-plane-operator/controllers/hostedcontrolplane/kas/certs.go

```diff
@@ -0,0 +1,163 @@
+package kas
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"net"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/openshift/hypershift/certs"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/pki"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/util"
+)
+
+const (
+	// Service signer secret keys
+	ServiceSignerPrivateKey = "service-account.key"
+	ServiceSignerPublicKey  = "service-account.pub"
+)
+
+func (p *KubeAPIServerParams) ReconcileServerCertSecret(secret, ca *corev1.Secret) error {
```

Comment:
My main piece of feedback that applies across the board here is that I think we'd be well served to make **all** of the reconcile functions 100% stateless. With the reconcile functions expressed as methods of parameter structs, the reconcile functions now have implicit state — there's no way to know (as an observer or test author) what the complete set of inputs are without mentally running the code (or doing some bespoke static analysis?) to find references to fields of `p`, which becomes kind of a god-object or global in the context of reconcile. In my experience this makes testing and comprehension extremely difficult.

My suggestion would be to make all the functions stateless, and if the parameter list is annoyingly long, consider introducing a tightly scoped input structure for that reconciliation task which can be populated from the caller to preserve statelessness and make it crystal clear what the inputs are.

Because of the enormity of this refactor I'm okay to merge what's here, but would strongly prefer we do a followup very soon to convert the new functions to be stateless, and also make it clear that new conversions to this style (and net net code) follow a stateless paradigm as a matter of stated design principle within the operator. Maybe some big TODOs at the top of these new files about this discussion would be warranted.

Thanks for taking on this work, it has been a long time coming!